src/compiler/checker.ts

```diff
@@ -38841,11 +38841,17 @@ namespace ts {
             if (!checkGrammarDecoratorsAndModifiers(node) && hasEffectiveModifiers(node)) {
                 grammarErrorOnFirstToken(node, Diagnostics.An_export_assignment_cannot_have_modifiers);
             }
+
+            const typeTag = isInJSFile(node) ? getJSDocTypeTag(node) : undefined;
+            if (typeTag) {
+                checkAssertionWorker(typeTag.typeExpression.type, typeTag.typeExpression.type, node.expression);

```

Comment:
I think this annotation should be treated like a declaration, not a cast. Specifically, that means the assignability is stricter.